### PR TITLE
Add Module-level Adapters, Save-Restore and tests

### PR DESCRIPTION
**Pull request by titu1994**
# What does this PR do ?

Adds significant functionality to adapters, allowing multiple module-specific adapters, save and restore of just the adapters themselves rather than the full model (saving several hundred megabytes-to-gigabytes), and a suite of tests for the new functionality.

**Collection**: [Core, ASR]

# Changelog 
- Adds support for both "global" and "module" specific adapters. Module specific adapters allow for targetted adapters to certain modules, independent of the others.
- Adds support for saving and restoring just the adapter modules, independent of the original model weights. This allows savings of significant disk storage, since just the adapters are usually a tiny fraction of the param count of the models.
- Adds a battery of tests for the new capabilities.
- Add support for dropout to adapter modules
- Add support for stochastic dropout to all adapters via ResidualAddStrategy.

# Usage

Subclass implementations can chose between global adapters, module adapters or both. A basic implementation of Adapters for a toy model is provided in the `tests/core/mixins/adapters/test_adapter_model_mixin.py`

```python
# Global adapters (adds "abc" adapter to all supported modules - encoder+decoder+others1)
model.add_adapter(name="abc", cfg=cfg) 

# Module adapters 
model.add_adapter(name="decoder:abc", cfg=cfg)

# Save all of the adapters (only the adapters themselves)
model.save_adapters("adapters.pt", name=None)
OR
model.save_adapters("adapters.pt", name={global or module adapter name})

# Restore one-or-all adapters (only the adapters themselves)
new_model.load_adapters("adapters.pt", name=None)
OR
new_model.load_adapters("adapters.pt", name={exact module name from state dict}, 
                        map_location='cpu', strict=True)
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

### CODE REVIEW

File: examples/asr/asr_adapters/train_asr_adapter.py [hunk 1]
```diff
@@ -168,6 +175,9 @@ def main(cfg):
     # Then, Unfreeze just the adapter weights that were enabled above (no part of encoder/decoder/joint/etc)
     model.unfreeze_enabled_adapters()
 
+    # Update model config prior to training
+    model.cfg = model.cfg
+
     # Finally, train model
     trainer.fit(model)
```
Comment (sam1373) [inline examples/asr/asr_adapters/train_asr_adapter.py:182]: Maybe add adapter-only saving/loading to this file

File: nemo/core/classes/modelPT.py [hunk 2]
```diff
@@ -1301,6 +1301,7 @@ def cfg(self):
             Changes to this config are not reflected in the state of the model.
             Please create a new model using an updated config to properly update the model.
         """
+        self._set_hparams(OmegaConf.create({'cfg': self._cfg}))
```
[the remark this answers is not in the document]
Comment (titu1994) [reply nemo/core/classes/modelPT.py:1304]: In order to always save the config. Otherwise the explicit `model.cfg = model.cfg` is required in order to preserve the latest config (well it may still be required, but best to preserve the config value during both get and set).

File: examples/asr/asr_adapters/train_asr_adapter.py [hunk 3]
```diff
@@ -142,8 +142,15 @@ def main(cfg):
 
     # Setup optimizer
     cfg.model.optim = update_model_cfg(model.cfg.optim, cfg.model.optim)
+    model.cfg.optim = cfg.model.optim  # Update the model's config
     model.setup_optimization(cfg.model.optim)
```
[the remark this answers is not in the document]
Comment (titu1994) [reply examples/asr/asr_adapters/train_asr_adapter.py:146]: Thats what the update_model_cfg() does - it looks up the old config inside the model, and merges the config with the values inside the user defined overrides. It wont allow deletion or addition but only override.

If we want to support addition and deletion then we have cannot use the internal config at all. We are only left with the user having to explicitly define everything or nothing.

File: examples/asr/asr_adapters/train_asr_adapter.py (hunk 1 above)
Comment (titu1994) [reply examples/asr/asr_adapters/train_asr_adapter.py:182]: Good point, forgot about that.

Comment (titu1994) [reply examples/asr/asr_adapters/train_asr_adapter.py:184]: Done

File: nemo/core/classes/modelPT.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sam1373) [reply nemo/core/classes/modelPT.py:1304]: It seems very strange to do this kind of update during a get.. And if we do, then why do we also still need that line? Having to do model.cfg = model.cfg is very unintuitive. Can we replace that with an explicit update_hparams method?

File: examples/asr/asr_adapters/train_asr_adapter.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sam1373) [reply examples/asr/asr_adapters/train_asr_adapter.py:147]: What I am saying is that by default I think defining those fully in the config makes more sense, as it would allow completely different classes for those, which would be useful. For a specific example, the default models likely use noam, which is not ideal if we're only doing a few adapter training steps; but for other optimizers we'd need to have different parameters, such as removing d_model from optimizer config. If we do still really want to keep the config merge functionality, perhaps we can have a special flag for overriding configs.

[the remark this answers is not in the document]
Comment (sam1373) [reply examples/asr/asr_adapters/train_asr_adapter.py:147]: In fact, thinking also about potentially applying adapters to ssl models, it would be even more important there to be able to completely re-do these configs, as things such as masking etc. may be completely different

[the remark this answers is not in the document]
Comment (titu1994) [reply examples/asr/asr_adapters/train_asr_adapter.py:147]: Hmm valid points. Ok we will go the "complete replacement approach" then.

File: nemo/core/classes/modelPT.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (titu1994) [reply nemo/core/classes/modelPT.py:1304]: Having yet another method to update configs seems redundant - we have the parameter in the class managing this. Sure it seems unintuitive but that's because there's no feasible way to 1) Give users a pointer to the config 2) let them modify in place 3) Update hyper parameters for wandb. 2 and 3 contradict each other.

We could take other approaches, as in explicitly calling this method inside of our "pre" train begin or whatever, but that would be outside the scope of this pr.

File: examples/asr/asr_adapters/train_asr_adapter.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (titu1994) [reply examples/asr/asr_adapters/train_asr_adapter.py:145]: Done

File: examples/asr/asr_adapters/train_asr_adapter.py [hunk 4]
```diff
@@ -120,7 +121,7 @@ def main(cfg):
         raise ValueError("Cannot set `cfg.model.nemo_model` and `cfg.model.pretrained_model`. Select one only.")
```
Comment (erastorgueva-nv) [inline examples/asr/asr_adapters/train_asr_adapter.py:121]: I'd write "Cannot set both `cfg.model.nemo_model` and `cfg.model.pretrained_model`. Set one only."

File: examples/asr/asr_adapters/train_asr_adapter.py [hunk 5]
```diff
@@ -168,9 +181,23 @@ def main(cfg):
     # Then, Unfreeze just the adapter weights that were enabled above (no part of encoder/decoder/joint/etc)
     model.unfreeze_enabled_adapters()
 
+    # Update model config prior to training
+    model.cfg = model.cfg
```
Comment (erastorgueva-nv) [inline examples/asr/asr_adapters/train_asr_adapter.py:185]: I don't understand what does this line do?

File: nemo/collections/asr/parts/mixins/asr_adapter_mixins.py [hunk 6]
```diff
@@ -167,3 +159,37 @@ def _check_valid_model_with_adapter_support(self):
...
+        Args:
+            name: A global adapter, or a module adapter name (with structure module_name:adapter_name).
+
+        Returns:
+            A tuple representing (module_name, adapter_name). If a global adapter is provided,
+            module_name is set to ''.
+        """
+        module_name, adapter_name = super().resolve_adapter_module_name_(name)
+
+        # resolve name and module onlt for valid modules
```
Comment (erastorgueva-nv) [inline nemo/collections/asr/parts/mixins/asr_adapter_mixins.py:181]: Typo: "only"

File: examples/asr/asr_adapters/train_asr_adapter.py (hunk 5 above)
Comment (titu1994) [reply examples/asr/asr_adapters/train_asr_adapter.py:185]: Its a copy of the internal config to the PTL cache, so that WandB will show the updated config in logs. Without this, configs will not update. Its somewhat a limitation of PTLs internal wandb log handling step.

**Review by sam1373 — APPROVED** [review]
